examples/sugarscape_ig/ss_polars/agents.py

```diff
@@ -25,77 +28,182 @@ def __init__(
-            pl.col("agent_order") >= pl.col("blocking_agent_order")
+            (pl.col("agent_order") >= pl.col("blocking_agent_order"))
+            | pl.col("blocking_agent_order").is_null()
         )
 
-        # Sort cells by sugar and radius (nearest first)
-        neighborhood = neighborhood.sort(["sugar", "radius"], descending=[True, False])
+        # Sort neighborhood by agent_order & max_sugar (max_sugar because we will check anyway if the cell is empty)
+        # However, we need to make sure that the current agent cell is ordered by current sugar (since it's 0 until agent hasn't moved)
+        neighborhood = neighborhood.with_columns(
+            max_sugar=pl.when(pl.col("blocking_agent_id") == pl.col("agent_id_center"))
+            .then(pl.lit(0))
+            .otherwise(pl.col("max_sugar"))
+        ).sort(
+            ["agent_order", "max_sugar", "radius", "dim_0"],
+            descending=[False, True, False, False],
+        )
+        return neighborhood
+
+    def get_best_moves(self, neighborhood: pl.DataFrame) -> pl.DataFrame:
+        """Get the best moves for each agent
+
+        Parameters
+        ----------
+        neighborhood : pl.DataFrame
+            Neighborhood DataFrame
 
+        Returns
+        -------
+        pl.DataFrame
+            DataFrame with the best moves for each agent
+        """
+        raise NotImplementedError("Subclasses must implement this method")
+
+
+class AntPolarsLoopDF(AntPolarsBase):
+    def get_best_moves(self, neighborhood: pl.DataFrame):
         best_moves = pl.DataFrame()
+
         # While there are agents that do not have a best move, keep looking for one
         while len(best_moves) < len(self.agents):
-            # Get the best moves for each agent and if duplicates are found, select the one with the highest order
+            # Check if there are previous agents that might make the same move (priority for the given move is > 1)
+            neighborhood = neighborhood.with_columns(
+                priority=pl.col("agent_order").cum_count().over(["dim_0", "dim_1"])
+            )
+
+            # Get the best moves for each agent:
+            # If duplicates are found, select the one with the highest order
             new_best_moves = (
                 neighborhood.group_by("agent_id_center", maintain_order=True)
                 .first()
-                .sort("agent_order")
-                .unique(subset=["dim_0", "dim_1"], keep="first")
+                .unique(subset=["dim_0", "dim_1"], keep="first", maintain_order=True)
             )
-
             # Agents can make the move if:
             # - There is no blocking agent
             # - The agent is in its own cell
             # - The blocking agent has moved before him
-            condition = pl.col("agent_id").is_null() | (
-                pl.col("agent_id") == pl.col("agent_id_center")
+            # - There isn't a higher priority agent that might make the same move
```

Comment:
How did you find out about the necessity of adding this condition? Useful for best practice / how-to guide on how this was overlooked previously.

Reply:
I found that out when testing with many potential collisions. 
The problem was the following:
1. Consider two agents targeting the same cell:
   - A mid-priority agent (higher in the agent order)
   - A low-priority agent (lower in the agent order)
2. The mid-priority agent has low preference for the cell
3. The low-priority agent has high preference for the cell
4. Without accounting for priority:
   - The mid-priority agent's best moves kept getting "stolen" by higher priority agents
   - This forced it to resort to lower preference target cells
   - However, these lower preference cells were often already taken by lower priority agents in previous iterations

The best-practice to prevent these race conditions, is to implement a "priority" count to ensure that each action is "legal". 

The downside is that the "priority" needs to be recomputed at each iteration and I think that's why is slower than Numba now. After the Ibis refactoring we can check if lazy evaluation can help mitigate the issue.

Reply:
Sounds messy. I hope it is possible to abstract out the move process, so that the user only has to provide the function on how the agent chooses the optimal location to move to.

Reply:
I believe it can definitely can be abstracted away in the implementation. See #108